### PR TITLE
[sdk-metrics] Fix race condition for MemoryPoint Reclaim

### PR DESCRIPTION
**Pull request by utpilla**
#### On the main branch:

An update thread does the following:
1. Get index for the `MetricPoint` array
2. Increment the `ReferenceCount` for the `MetricPoint` at that index
3.  Ensure the `MetricPoint` is valid for use
4. Update necessary values
5. Set the `MetricPointStatus` to `CollectPending`
6. Decrement the `ReferenceCount` for the `MetricPoint`

The collect thread does the following:
1. Enumerate over all the MetricPoints
2. If the `MetricPointStatus` for a given `MetricPoint` is `NoCollectPending`, then check if it can be marked invalid
4. Try marking the `MetricPoint` invalid (this happens by setting the `ReferenceCount` to `int.MinValue` when no one is using it)
5. If the 3rd step was successful, then reclaim the `MetricPoint`

#### Where's the race condition?

Consider this sequence of steps where both the Update thread and the Collect thread are working on the same MetricPoint

| Time | Update thread |  Collect thread |
|-----  |-----------------|------------------|
|    T1 | Get `MetricPoint` index        |           Enumerate over all `MetricPoints`                |
|    T2 |  Update thread gets switched out     |    Check `MetricPointStatus` for the `MetricPoint`                      |
|    T3 | Update thread is switched out      |   Collect thread finds that `MetricPointStatus` is `NoCollectPending`                      |
|    T4 | Update thread is back and it increments the `ReferenceCount` to `1`        |  Collect thread gets switched out              |
|    T5 | Ensure that the `MetricPoint` is valid for use        |   Switched out                         |
|    T6 | Update the required values          |                           Switched out |
|    T7 | Set `MetricPointStatus` to `CollectPending`  **(This is the update that we would miss)**       |   Switched out                        |
|    T8 | Decrement the `ReferenceCount` to `0`         |    Switched out                       |
|    T9 |        |   Collect thread is back and sets the `ReferenceCount` to `int.MinValue` as it finds the `ReferenceCount` to be `0`                     |
|    T10 |       |  Reclaims the `MetricPoint` and misses the update that happened at T7                      |

#### Fix

I'm introducing a double-checked locking type construct to recheck if the `MetricPointStatus` was changed to `CollectPending` before the Collect thread could mark the MetricPoint invalid for use. When that happens, we would now call `Snapshot` for that `MetricPoint` and mark it to be reclaimed in the next Collect cycle. Note that the `MetricPoint` would remain invalid for use until the next Collect cycle reclaims it.

### CODE REVIEW

**Comment by CodeBlanch** [comment]
@Yun-Ting Would you mind spinning up an issue to make sure we have coyote test coverage over this delta metric point reclaim feature?

File: src/OpenTelemetry/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,10 @@
   function when configuring a view (applies to individual metrics).
   ([#5542](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5542))
 
+* Fixed a race condition for MetricPoint reclaim scenario which could have led
+  to a measurement being dropped.
+  ([#5546](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5546))
```
Comment (CodeBlanch) [inline src/OpenTelemetry/CHANGELOG.md:12]: I'm worried the impact isn't super clear and we might cause some panic. Could we somehow indicate it is an opt-in thing? Like...

* Fixed a race condition for MetricPoint reclaim scenario (enabled via `OTEL_DOTNET_EXPERIMENTAL_METRICS_RECLAIM_UNUSED_METRIC_POINTS`) which could have led
  to a measurement being dropped.
  ([#5546](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5546))

Comment (utpilla) [reply src/OpenTelemetry/CHANGELOG.md]: Updated.

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**